pkg/envoy/resources.go

```diff
@@ -53,8 +55,18 @@ var (
 	// NetworkPolicyHosts. Resources in this cache must have the
 	// NetworkPolicyHostsTypeURL type URL.
 	NetworkPolicyHostsCache = newNPHDSCache()
+
+	observerOnce = sync.Once{}
 )
 
+// HandleResourceVersionAck is required to implement ResourceVersionAckObserver.
+func (cache *NPHDSCache) HandleResourceVersionAck(ackVersion uint64, nackVersion uint64, node *envoy_api_v2_core.Node, resourceNames []string, typeURL string, detail string) {
+	// Start caching for IP/ID mappings on the first indication someone wants them
+	observerOnce.Do(func() {
+		ipcache.IPIdentityCache.AddListener(cache)
```

Comment:
The commit message says that this is done when:

> the first xDS request for NetworkPolicyHosts is received.

But this seems to be when the first ack is received (`HandleResourceVersionAck()`). Is this just a naming convention, is this effectively the request from the proxy?

Reply:
Each xDS request carries an ack/nack indication of what has been succesfully received or actively NACKed so far. The first request carries empty (strings), but the `HandleResourceVersionAck` is still called. Also, if cilium restarts and (a sidecar) proxy keeps running, the first xDS request received by the cilium-agent will carry an actual ack, but the effect here will be the same. Previously we did not process acks/nacks for NPHDS at all. Now we added the cache itself as the ACK observer via `HandleResourceVersionAck` which will be called for each request (passing some basic sanity filtering) and we will simply bootstrap the identity caching at the first of those calls.

Maybe I should have only said that the first request carries an ACK for the proxy being in a blank state :-)